<div style="text-align: center; background-color: #0A6EBD; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Final project - Introduction to DS
</div>

<div style="text-align: center; background-color: #b1d1ff; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Stage 02 - Data preprocessing & Basic exploration
</div>

## Import

In [1]:
import os
import sys
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
import isodate

## Read raw data from file

In [2]:
raw_df = pd.read_csv("../data/raw/video_data_raw.csv")
raw_df.sample(n=5)

video_id                               channelTitle  \
15186  lokv-h480YI                             Great Learning   
56357  nzmc_1Ys854  The TWIML AI Podcast with Sam Charrington   
39029  bYClWlt8DjQ                                    MyStudy   
3108   cXDpuigSsjg                                    sentdex   
41088  OvFJjqVl--0                                  itversity   

                                                   title  \
15186  Clustering in Data mining | K means Clustering...   
56357  Human-Centered ML for High-Risk Behaviors with...   
39029               Deep Learning  Brain vs Computer - 1   
3108   [See Description] Pipeline Tutorial - Python f...   
41088  DBT Workshop - Data Transformations using DBT ...   

                                             description  \
15186  🔥1000+ Free Courses With Free Certificates: ht...   
56357  Today we’re joined by Stevie Chancellor, an As...   
39029  This tutorial will explain basic difference an...   
3108   UPDATED series: https://pythonprogramming.net/...   
41088  As part of this live workshop we will walk you...   

                                                    tags  \
15186  ['Great Learning', 'Great Lakes', 'Data Analyt...   
56357  ['TWiML & AI', 'Podcast', 'Tech', 'Technology'...   
39029  ['brain vs deep learning', 'deep learning vs b...   
3108   ['finance', 'algorithmic trading', 'automated'...   
41088  ['DBT', 'CLI', 'DBT Cloud', 'EMR Cluster', 'Th...   

                publishedAt  viewCount  likeCount  favouriteCount  \
15186  2020-08-05T10:30:53Z     8130.0      182.0             NaN   
56357  2021-04-15T15:56:59Z      267.0        5.0             NaN   
39029  2016-10-13T05:05:58Z      512.0        4.0             NaN   
3108   2016-03-04T16:56:06Z    11778.0       73.0             NaN   
41088  2022-08-27T02:34:34Z     6075.0       94.0             NaN   

       commentCount   duration definition  caption  
15186          12.0  PT3H37M9S         hd    False  
56357           0.0    PT42M8S         hd    False  
39029           0.0    PT8M54S         hd    False  
3108           16.0   PT15M23S         hd    False  
41088           8.0  PT1H55M4S         hd    False

### 📌 How many rows and how many columns does the raw data have?

In [3]:
data_video_shape = raw_df.shape
print(f"Video data current shape: {data_video_shape}")

Video data current shape: (60032, 13)


Answer: Our raw video data has `60032 rows` and `13 columns`, which meets the given requirement: "Your dataset must have at least `5 fields` and `1000 observations`".

### 📌 What does each line mean? Does it matter if the lines have different meanings?

Answer: 
- Based on the observations of the dataset, we suppose that each line represents a unique set of details regarding individual YouTube videos. 
- Looking through the data table of the csv file, it seems there are no "out of place" lines.

### 📌 Does the raw data have duplicate rows?

In [4]:
# retrieve the index
index = raw_df.index
# create a Pandas Series indicating whether each index is duplicated or not
deDupSeries = index.duplicated(keep='first')
# calculate the number of duplicated rows
num_duplicated_rows = deDupSeries.sum()

In [5]:
if num_duplicated_rows == 0:
    print(f"Raw data have no duplicated line!")
else:
    ext = "lines" if num_duplicated_rows > 1 else "line"
    print(f"Raw data have {num_duplicated_rows} duplicated " + ext)

Raw data have no duplicated line!


### 📌 What does each column mean?

<table style="width: 100%; border-collapse: collapse;">
  <thead>
    <tr>
      <th style="min-width: 10px; width: 30px; background-color: #04B1CC; color: white; font-size :15px; font-weight: bold; text-align: left; padding: 8px;">No</th>
      <th style="background-color: #04B1CC; color: white; font-size :15px; font-weight: bold; text-align: left; padding: 8px;">Columns</th>
      <th style="background-color: #04B1CC; color: white; font-size :15px; font-weight: bold; text-align: left; padding: 8px;">Meaning</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="font-size: 14px; text-align: left;">1</td>
      <td style="font-size: 14px; text-align: left;">video_id</td>
      <td style="font-size: 14px; text-align: left;">Unique identifier for the YouTube video.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">2</td>
      <td style="font-size: 14px; text-align: left;">channelTitle</td>
      <td style="font-size: 14px; text-align: left;">The name of the channel that uploaded the video.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">3</td>
      <td style="font-size: 14px; text-align: left;">title</td>
      <td style="font-size: 14px; text-align: left;">Title of the video.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">4</td>
      <td style="font-size: 14px; text-align: left;">description</td>
      <td style="font-size: 14px; text-align: left;">Description or summary of the video content.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">5</td>
      <td style="font-size: 14px; text-align: left;">tags</td>
      <td style="font-size: 14px; text-align: left;">Keywords or tags associated with the video.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">6</td>
      <td style="font-size: 14px; text-align: left;">publishedAt</td>
      <td style="font-size: 14px; text-align: left;">Date and time when the video was published.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">7</td>
      <td style="font-size: 14px; text-align: left;">viewCount</td>
      <td style="font-size: 14px; text-align: left;">Number of views the video has accumulated.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">8</td>
      <td style="font-size: 14px; text-align: left;">likeCount</td>
      <td style="font-size: 14px; text-align: left;">Count of likes received by the video.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">9</td>
      <td style="font-size: 14px; text-align: left;">favoriteCount</td>
      <td style="font-size: 14px; text-align: left;">Deprecated; used to track how many times viewers added the video to their favorites.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">10</td>
      <td style="font-size: 14px; text-align: left;">commentCount</td>
      <td style="font-size: 14px; text-align: left;">Number of comments posted on the video.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">11</td>
      <td style="font-size: 14px; text-align: left;">duration</td>
      <td style="font-size: 14px; text-align: left;">Length of the video.</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">12</td>
      <td style="font-size: 14px; text-align: left;">definition</td>
      <td style="font-size: 14px; text-align: left;">Video resolution or quality (e.g., HD, SD).</td>
    </tr>
    <tr>
      <td style="font-size: 14px; text-align: left;">13</td>
      <td style="font-size: 14px; text-align: left;">caption</td>
      <td style="font-size: 14px; text-align: left;"> Indicates whether closed captions are available for the video.</td>
    </tr>
  </tbody>
</table>

❓ TODO: Liệu có thông tin nào chỉ đúng ở thời điểm dữ liệu được thu thập, sau đó nó có thể bị thay đổi
vd: count sẽ tăng lên/giảm xuống (lượt like); 

### 📌 What data type does each column currently have? Are there any columns whose data types are not suitable for further processing?

In [6]:
raw_df.dtypes

video_id           object
channelTitle       object
title              object
description        object
tags               object
publishedAt        object
viewCount         float64
likeCount         float64
favouriteCount    float64
commentCount      float64
duration           object
definition         object
caption              bool
dtype: object

🔨 We notice that the columns `publishedAt` and `duration` are currently of object type. Given that they represent time periods, it is advisable to convert `publishedAt`  to datetime type. As the `duration` is in **ISO 8601** format, we'll convert it into an `integer` representing the total number of seconds

In [7]:
# convert publishedAt to datetime
raw_df["publishedAt"] = pd.to_datetime(raw_df["publishedAt"])
# convert duration to integer
raw_df['duration'] = raw_df['duration'].apply(lambda x: isodate.parse_duration(x))
raw_df['duration'] = raw_df['duration'].dt.total_seconds()
raw_df['duration'] = raw_df['duration'].astype(np.int64)

✨ Since the `publishedAt` column only carries general values, we can extract more useful information from this data. The special thing is that the extracted information will be `categorical` values.

In [8]:
if 'publishedAt' in raw_df.columns:
    # Extract time features from `publishedAt` column
    raw_df['publishedAt.hour']       = raw_df['publishedAt'].dt.hour
    raw_df['publishedAt.day']        = raw_df['publishedAt'].dt.day
    raw_df['publishedAt.year']       = raw_df['publishedAt'].dt.year
    raw_df['publishedAt.month']      = raw_df['publishedAt'].dt.month
    raw_df['publishedAt.dayOfWeek']  = raw_df['publishedAt'].dt.dayofweek  # Monday = 0, Sunday = 6
    raw_df['publishedAt.weekOfYear'] = raw_df['publishedAt'].dt.isocalendar().week

    # Change datatype of the above columns to categorical
    raw_df['publishedAt.hour']       = raw_df['publishedAt.hour'].astype('str')
    raw_df['publishedAt.day']        = raw_df['publishedAt.day'].astype('str')
    raw_df['publishedAt.year']       = raw_df['publishedAt.year'].astype('str')
    raw_df['publishedAt.month']      = raw_df['publishedAt.month'].astype('str')
    raw_df['publishedAt.dayOfWeek']  = raw_df['publishedAt.dayOfWeek'].astype('str')
    raw_df['publishedAt.weekOfYear'] = raw_df['publishedAt.weekOfYear'].astype('str')

In [9]:
# TEST
raw_df.dtypes

video_id                               object
channelTitle                           object
title                                  object
description                            object
tags                                   object
publishedAt               datetime64[ns, UTC]
viewCount                             float64
likeCount                             float64
favouriteCount                        float64
commentCount                          float64
duration                                int64
definition                             object
caption                                  bool
publishedAt.hour                       object
publishedAt.day                        object
publishedAt.year                       object
publishedAt.month                      object
publishedAt.dayOfWeek                  object
publishedAt.weekOfYear                 object
dtype: object

### 📌 With each numerical column, how are values distributed?

📚 For columns with numeric data types, we will calculate some basic statistics:
- Percentage (from 0 to 100) of missing values
- The min
- The lower quartile (phân vị 25)
- The median (phân vị 50)
- The upper quartile (phân vị 75)
- The max

Then we will observe and comment to see if the results are unusual?

In [10]:
# YOUR CODE HERE
num_col_info_df = raw_df.select_dtypes(exclude=['object', 'bool'])

def missing_ratio(s):
    return (s.isna().mean() * 100)

def median(df):
    return (df.quantile(0.5))

def lower_quartile(df):
    return (df.quantile(0.25))

def upper_quartile(df):
    return (df.quantile(0.75))

num_col_info_df = num_col_info_df.agg([missing_ratio, "min", lower_quartile, median, upper_quartile, "max"])
num_col_info_df

publishedAt     viewCount      likeCount  \
missing_ratio                                0.0  6.663113e-03       0.404784   
min                    2006-10-25 10:28:09+00:00  0.000000e+00       0.000000   
lower_quartile         2019-02-18 04:02:18+00:00  6.220000e+02      11.000000   
median          2020-12-22 09:26:05.500000+00:00  3.059500e+03      55.000000   
upper_quartile  2022-07-05 06:04:02.500000+00:00  1.602175e+04     334.000000   
max                    2023-11-24 13:04:35+00:00  3.447645e+07  571358.000000   

                favouriteCount  commentCount  duration  
missing_ratio            100.0      1.279318       0.0  
min                        NaN      0.000000       0.0  
lower_quartile             NaN      0.000000     293.0  
median                     NaN      5.000000     695.0  
upper_quartile             NaN     27.000000    1786.0  
max                        NaN  60054.000000   92218.0

Observation: We notice that some columns have missing data and need to be resolved before moving onto the next analysis steps. Other than that, the data does not appear to be unusual.

### 🚧 Preprocessing numerical columns

We notice that `favouriteCount` does not have any value, so we can drop this feature

In [11]:
# drop favouriteCount
if "favouriteCount" in raw_df.columns:
    raw_df = raw_df.drop('favouriteCount', axis=1)

`viewCount`, `likeCount` and `commentCount` have little missing value, we'll fill them with their `median` and then convert them to integer

In [12]:
# fill missing value with median
missing_cols = ['viewCount', 'likeCount', 'commentCount']
for col in missing_cols:
    raw_df[col] = raw_df[col].fillna(raw_df[col].median().__round__(0))
    raw_df[col] = raw_df[col].astype(np.int64)

In [13]:
# TEST
raw_df.select_dtypes(exclude=['object', 'bool'])\
    .agg([missing_ratio, "min", lower_quartile, median, upper_quartile, "max"])

publishedAt    viewCount  likeCount  \
missing_ratio                                0.0         0.00        0.0   
min                    2006-10-25 10:28:09+00:00         0.00        0.0   
lower_quartile         2019-02-18 04:02:18+00:00       622.00       11.0   
median          2020-12-22 09:26:05.500000+00:00      3060.00       55.0   
upper_quartile  2022-07-05 06:04:02.500000+00:00     16017.25      332.0   
max                    2023-11-24 13:04:35+00:00  34476453.00   571358.0   

                commentCount  duration  
missing_ratio            0.0       0.0  
min                      0.0       0.0  
lower_quartile           1.0     293.0  
median                   5.0     695.0  
upper_quartile          26.0    1786.0  
max                  60054.0   92218.0

### 📌 With each categorical column, how are values distributed?

📚 For columns with non-numeric data types, we calculate:
- Percentage (from 0 to 100) of missing values
- Number of different values (and we do not consider missing values)
- Show a few values and percentage (from 0 to 100) of each value sorted by decreasing percentage (we do not consider missing values, the ratio is the ratio compared to the number of non-missing values)

Then we will observe and comment to see if the results are unusual?

In [14]:
pd.set_option('display.max_colwidth', 100) # For clearly
pd.set_option('display.max_columns', None) # For clearly

In [15]:
cat_col_info_df = raw_df.select_dtypes(include=['object', 'bool'])

def missing_ratio(s):
    return (s.isna().mean() * 100)

def num_values(s):
    s = s.astype('str').str.split(';')
    s = s.explode()
    return len(s.value_counts())

def value_ratios(s):
    s = s.astype('str').str.split(';')
    s = s.explode()
    totalCount = (~s.isna()).sum()
    return ((s.value_counts()/totalCount*100).round(1)).to_dict()

cat_col_info_df = cat_col_info_df.agg([missing_ratio, num_values, value_ratios])
cat_col_info_df

video_id  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                   60032   
value_ratios   {'WkqM0ndr42c': 0.0, 'ls8hEDjhKU0': 0.0, 'Fh0ArGT2_b0': 0.0, '_tHCoU5TZZg': 0.0, 'fXAvXLOUjlU': ...   

                                                                                                      channelTitle  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                     160   
value_ratios   {'itversity': 5.8, 'Databricks': 4.7, 'Great Learning': 3.1, 'Analytics India Magazine': 3.0, 'K...   

                                                                                                             title  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                   59644   
value_ratios   {'TL': 0.1, 'Data Analyst MENTORSHIP -  Q&A (while I drink coffee)': 0.1, 'Machine Learning Summ...   

                                                                                                       description  \
missing_ratio                                                                                             2.803505   
num_values                                                                                                   53351   
value_ratios   {'nan': 2.7, 'Connect with me or follow me at
https://www.linkedin.com/in/durga0gadiraju
https...   

                                                                                                              tags  \
missing_ratio                                                                                            18.956557   
num_values                                                                                                   35545   
value_ratios   {'nan': 18.9, '['Databricks']': 2.2, '['CBMM', 'Center for Brains Minds and Machines', 'Artifici...   

                            definition                        caption  \
missing_ratio                      0.0                            0.0   
num_values                           2                              2   
value_ratios   {'hd': 95.5, 'sd': 4.5}  {'False': 88.6, 'True': 11.4}   

                                                                                                  publishedAt.hour  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                      24   
value_ratios   {'15': 8.6, '13': 8.5, '16': 7.8, '14': 7.6, '12': 6.3, '17': 5.3, '4': 4.5, '11': 4.4, '18': 3....   

                                                                                                   publishedAt.day  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                      31   
value_ratios   {'6': 3.9, '25': 3.7, '7': 3.6, '28': 3.6, '11': 3.6, '19': 3.5, '14': 3.5, '21': 3.5, '26': 3.4...   

                                                                                                  publishedAt.year  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                      17   
value_ratios   {'2020': 17.5, '2021': 16.6, '2023': 16.6, '2022': 16.2, '2019': 9.7, '2018': 7.2, '2017': 6.1, ...   

                               

Observation: We notice that some columns have missing data and need to be resolved before moving onto the next analysis steps. Other than that, the data does not appear to be unusual.

### 🚧 Preprocessing categorical columns

#### Column `description`

Let's observe some values ​​in the `description` column

In [16]:
raw_df['description'].sample(n=5)

31062    #DataTrained #ChatGPT #IntroductionofChatGPT \r\n\r\nAre you curious about how ChatGPT can revol...
55553    ⭐ Subscribe to Hindi Channel - https://www.youtube.com/@datawalebhaiya219\r\n🔴 Register for Code...
26510    Sravan Kasarla, CDO of Thrivent, outlines how the financial services provider is building a mode...
21908    In this episode of Student Success Stories, Karen Peterson talks about her experience in Springb...
3251     This Python programming tutorial covers how to pickle objects in Python. \r\n\r\nThe pickle modu...
Name: description, dtype: object

Việc tiền xử lý các giá trị bị thiếu trong cột này tương đối phức tạp vì nó yêu cầu chúng ta phải tạo ra một đoạn văn bản. Để cho đơn giản thì ta có thể điền giá trị `(nodescription)` vào các ô trống

In [17]:
raw_df['description'] = raw_df['description'].fillna("(nodescription)")

#### Column `tags`

Let's observe some values ​​in the `tags` column

In [18]:
raw_df['tags'].sample(n=5)

11914    ['brandon foltz', 'statistics 101', 'nominal data', 'introductory statistics', 'statistics', 'an...
45986                                                                     ['ML Application Into Production']
55491    ['shashank mishra', 'e learning bridge', 'data science', 'data scientist', 'a day in life at goo...
38703    ['Support Vector Machines', 'SVM in machine learning', 'svm kernel', 'used for classification', ...
30231                                      ['pix2pix tutorial', 'pix2pix pytorch', 'pix2pix implementation']
Name: tags, dtype: object

The `tags` column has a "special" data storage format: it is a `str` but has a structure similar to a `list` containing `str` elements inside. This can be "confusing" for other team members, so we will change the format of the `tags` string as follows: "`str_1|str_2[...]`"

In [19]:
raw_df['tags'] = raw_df['tags'].apply(lambda x: '|'.join(x.strip("[]").replace("\'", "").split(', ')) if isinstance(x, str) else x)

# Display the result
raw_df['tags'].sample(n=5)

37447    data analyst hierarchy|data analyst career path|is management good path for data analyst|is indi...
59252                                     data mining|text mining|machine learning|text analysis|import text
4396     Building Robust Machine Learning Models|Feature engineering|Machine Learning|Model Engineering|M...
15983                                                         Data Science|Data Scientist|Learn Data Science
55356                       quick insights|donald knuth|advice|good science|popular science|computer science
Name: tags, dtype: object

Cột `tags` vẫn còn tình trạng thiếu dữ liệu, việc thiếu dữ liệu này có lẽ xuất phát từ việc người đăng không thêm tags vào video của họ, ngoài ra việc dự đoán giá trị cho các video thiếu dữ liệu cũng tương đối phức tạp nên ta sẽ điền giá trị `(notag)` vào các vị trí trống

In [20]:
# Replace missing value with "(notag)"
raw_df['tags'] = raw_df['tags'].fillna("(notag)")

In [21]:
# TEST
raw_df.select_dtypes(include=['object', 'bool'])\
    .agg([missing_ratio, num_values, value_ratios])

video_id  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                   60032   
value_ratios   {'WkqM0ndr42c': 0.0, 'ls8hEDjhKU0': 0.0, 'Fh0ArGT2_b0': 0.0, '_tHCoU5TZZg': 0.0, 'fXAvXLOUjlU': ...   

                                                                                                      channelTitle  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                     160   
value_ratios   {'itversity': 5.8, 'Databricks': 4.7, 'Great Learning': 3.1, 'Analytics India Magazine': 3.0, 'K...   

                                                                                                             title  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                   59644   
value_ratios   {'TL': 0.1, 'Data Analyst MENTORSHIP -  Q&A (while I drink coffee)': 0.1, 'Machine Learning Summ...   

                                                                                                       description  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                   53351   
value_ratios   {'(nodescription)': 2.7, 'Connect with me or follow me at
https://www.linkedin.com/in/durga0gad...   

                                                                                                              tags  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                   35545   
value_ratios   {'(notag)': 18.9, 'Databricks': 2.2, 'CBMM|Center for Brains Minds and Machines|Artificial Intel...   

                            definition                        caption  \
missing_ratio                      0.0                            0.0   
num_values                           2                              2   
value_ratios   {'hd': 95.5, 'sd': 4.5}  {'False': 88.6, 'True': 11.4}   

                                                                                                  publishedAt.hour  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                      24   
value_ratios   {'15': 8.6, '13': 8.5, '16': 7.8, '14': 7.6, '12': 6.3, '17': 5.3, '4': 4.5, '11': 4.4, '18': 3....   

                                                                                                   publishedAt.day  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                      31   
value_ratios   {'6': 3.9, '25': 3.7, '7': 3.6, '28': 3.6, '11': 3.6, '19': 3.5, '14': 3.5, '21': 3.5, '26': 3.4...   

                                                                                                  publishedAt.year  \
missing_ratio                                                                                                  0.0   
num_values                                                                                                      17   
value_ratios   {'2020': 17.5, '2021': 16.6, '2023': 16.6, '2022': 16.2, '2019': 9.7, '2018': 7.2, '2017': 6.1, ...   

                              

### 💾 Save the processed data

In [22]:
print(f"Total number of features: {raw_df.shape[1]}")
raw_df.dtypes

Total number of features: 18


video_id                               object
channelTitle                           object
title                                  object
description                            object
tags                                   object
publishedAt               datetime64[ns, UTC]
viewCount                               int64
likeCount                               int64
commentCount                            int64
duration                                int64
definition                             object
caption                                  bool
publishedAt.hour                       object
publishedAt.day                        object
publishedAt.year                       object
publishedAt.month                      object
publishedAt.dayOfWeek                  object
publishedAt.weekOfYear                 object
dtype: object

In [23]:
# Save processed data to disk
raw_df.to_csv("../data/processed/" + "video_data_processed.csv", index=False)